### PR TITLE
#8482: Trace on remote chips

### PR DESCRIPTION
**Pull request by tt-asaigal**
### Ticket
- https://github.com/tenstorrent/tt-metal/issues/8482

### Problem description
- Trace is not working on remote chips.

### What's changed
 - Changes from @pgkeller:
     Fix host completion write ptr in test_prefetcher
     Was clobbering cmds.
     This enables more prefetcher test configs w/ exec_buf

     Rework split prefetch exec_buf sync mechanism
     prefetch_h now stalls itself at the start of exec_buf by taking away its
     downstream page credits.  After sending the exec_buf, it unstalls the
     prefetch_q so that it can continue to do prefetch work.  dispatch restores
     prefetch_h credits at exec_buf_end.

     Increase test_prefetcher exec_buf test coverage

 - cq_prefetch changes:
   - Modify detection of exec_buf command. Prefetcher no longer looks at the cmd_id, since exec_buf command is not the first command in the Fetch Q. Instead rely on stall flag.
   - prefetch_h should not offset exec_buf command by preamble size, since it belongs to the same entry as a previously offset command (due to barrier_and_stall)
   - Account for wrap in barrier and stall

 - Add T3K trace tests.

### Checklist
- [x] Post commit CI passes
- [ ] Model regression CI testing passes (if applicable)
- [x] New/Existing tests provide coverage for changes

### CODE REVIEW

File: tt_metal/impl/device/device.cpp [hunk 1]
```diff
@@ -559,13 +559,15 @@ void Device::update_workers_build_settings(std::vector<std::vector<std::tuple<tt
...
+                auto prefetch_h_settings = std::get<1>(device_worker_variants[PREFETCH][0]);
+                auto prefetch_physical_core = prefetch_h_settings.worker_physical_core
                 TT_ASSERT(num_dispatchers == demux_settings.semaphores.size(), "Demux does not have required number of semaphores for Dispatchers. Exptected = {}. Fount = {}", num_dispatchers, demux_settings.semaphores.size());
                 uint32_t demux_sem = demux_settings.producer_semaphore_id;
                 for (auto&[core, settings] : device_worker_variants[DISPATCH]) {
                     auto dispatch_core_type = settings.dispatch_core_type;
                     settings.upstream_cores.push_back(demux_settings.worker_physical_core);
                     settings.downstream_cores.push_back(tt_cxy_pair(0, 0, 0));
-                    settings.compile_args.resize(17);
+                    settings.compile_args.resize(21);
```
Comment (cfjchu) [inline tt_metal/impl/device/device.cpp:570]: what is this magic number?

File: tt_metal/impl/device/device.cpp [hunk 2]
```diff
@@ -582,8 +584,13 @@ void Device::update_workers_build_settings(std::vector<std::vector<std::tuple<tt
                     compile_args[12] = 0; // unused: local ds semaphore
                     compile_args[13] = 0; // unused: remote ds semaphore
                     compile_args[14] = 0; // preamble size
-                    compile_args[15] = false; // is_dram_variant
-                    compile_args[16] = true; // is_host_variant
+                    // HEREEEEE
```
Comment (cfjchu) [inline tt_metal/impl/device/device.cpp:587]: remove HEREEEEE :)

File: tt_metal/impl/device/device.cpp (hunk 1 above)
Comment (tt-asaigal) [reply tt_metal/impl/device/device.cpp:570]: This is the size of the preallocated compile args vector sent to to dispatch core. We had to add 4 more compile args, hence 17 -> 24.

File: tt_metal/impl/device/device.cpp (hunk 2 above)
Comment (tt-asaigal) [reply tt_metal/impl/device/device.cpp]: Its not here in the latest commit.

File: tt_metal/impl/device/device.cpp (hunk 1 above)
Comment (xanderchin) [reply tt_metal/impl/device/device.cpp:569]: Can we de magicify this?

File: tt_metal/impl/device/device.cpp [hunk 3]
```diff
@@ -370,7 +370,7 @@ void Device::update_workers_build_settings(std::vector<std::vector<std::tuple<tt
                     uint32_t downstream_cb_base = mux_settings.cb_start_address + mux_settings.cb_size_bytes * mux_sem;
                     settings.upstream_cores.push_back(tt_cxy_pair(0, 0, 0));
                     settings.downstream_cores.push_back(mux_settings.worker_physical_core);
-                    settings.compile_args.resize(23);
+                    settings.compile_args.resize(22);
```
Comment (xanderchin) [inline tt_metal/impl/device/device.cpp:373]: magic?

**Review by pgkeller — APPROVED** [review]

**Review by cfjchu — APPROVED** [review]

File: tests/ttnn/unit_tests/test_multi_device_trace.py [hunk 4]
```diff
@@ -8,26 +8,31 @@
...
-    if pcie_device_mesh.get_num_devices() <= 1:
+def test_multi_device_single_trace(t3k_device_mesh, shape, use_all_gather, enable_async):
+    if t3k_device_mesh.get_num_devices() <= 1:
         pytest.skip("This test requires multiple devices")
 
     # Trace requires program cache to be enabled
-    for device_id in pcie_device_mesh.get_device_ids():
-        pcie_device_mesh.get_device(device_id).enable_async(enable_async)
-        pcie_device_mesh.get_device(device_id).enable_program_cache()
+    for device_id in t3k_device_mesh.get_device_ids():
```
Comment (cfjchu) [inline tests/ttnn/unit_tests/test_multi_device_trace.py:29]: btw, you can do `for device in t3k_device_mesh.devices()`

**Comment by ttmchiou** [comment]
can you run the t3k frequent tests pipeline to test your changes?

**Review by ttmchiou — APPROVED** [review]

File: tests/ttnn/unit_tests/test_multi_device_trace.py (hunk 4 above)
Comment (tt-asaigal) [reply tests/ttnn/unit_tests/test_multi_device_trace.py:29]: Another trace pr coming soon with test. I'll add this change there. Thanks for the suggestion :)

**Comment by tt-asaigal** [comment]
> can you run the t3k frequent tests pipeline on your branch to test your changes? will approve if the specific test suite in t3k frequent CI pipeline tests pass

Done, trace stress tests were passing

**Pull request merged**